dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -0,0 +1,219 @@
+        switch (res)
+        {
+        case MPMCQueueResult::FINISHED:
+            // when finished, we still should pop the remaings.
+            return kickRemaingsOnce(data) ? MPMCQueueResult::OK : MPMCQueueResult::FINISHED;
+        case MPMCQueueResult::EMPTY:
+        {
+            // Double check if this queue is empty.
+            std::lock_guard lock(mu);
+            res = mpmc_queue.tryPop(data);
+            switch (res)
+            {
+            case MPMCQueueResult::FINISHED:
+            case MPMCQueueResult::EMPTY:
+                // when finished, we still should pop the remaings.
+                // when mpmc_queue empty, we can try pop from remaings.
+                return kickRemaingsOnceWithoutLock(data) ? MPMCQueueResult::OK : res;
+            case MPMCQueueResult::OK:
+                kickRemaingsWithoutLock();
+            default:
+                return res;
+            }
+        }
+        case MPMCQueueResult::OK:
+            kickRemaings();
+        default:
+            return res;
+        }
+    }
+
+    size_t size() const
+    {
+        size_t mpmc_queue_size = mpmc_queue.size();
+        std::lock_guard lock(mu);
+        return mpmc_queue_size + remaings.size();
+    }
+
+    // When the queue is finished, it may appear that the mpmc_queue is empty but the remaings are not, causing isFull to return true.
+    // But this is expected, and the return value of isFull is meaningless after finished/cancelled.
+    bool isFull() const
+    {
+        {
+            std::lock_guard lock(mu);
+            if (!remaings.empty())
+                return true;
+        }
+        return mpmc_queue.isFull();
+    }
+
+    MPMCQueueStatus getStatus() const
+    {
+        return mpmc_queue.getStatus();
+    }
+
+    /// Cancel a NORMAL queue will wake up all blocking readers and writers.
+    /// After `cancel()` the queue can't be pushed or popped any more.
+    /// That means some objects may leave at the queue without poped.
+    bool cancel()
+    {
+        return cancelWith("");
+    }
+    bool cancelWith(String reason)
+    {
+        if (mpmc_queue.cancelWith(std::move(reason)))
+        {
+            std::lock_guard lock(mu);
+            remaings.clear();
+            return true;
+        }
+        return false;
+    }
+
+    const String & getCancelReason() const
+    {
+        return mpmc_queue.getCancelReason();
+    }
+
+    /// Finish a NORMAL queue will wake up all blocking readers and writers.
+    /// After `finish()` the queue can't be pushed any more while `pop` is allowed
+    /// the queue is empty.
+    /// Return true if the previous status is NORMAL.
+    bool finish()
+    {
+        return mpmc_queue.finish();
+    }
+
+private:
+    void kickRemaingsWithoutLock()
+    {
+        while (!remaings.empty() && mpmc_queue.tryPush(std::move(remaings.back())) == MPMCQueueResult::OK)
+            remaings.pop_back();
+    }
+
+    void kickRemaings()
```

Comment:
This function is intended for pushing `remains` to queue. So I suggest changing the name to `pushRemains`.

Reply:
ok, renamed.